router/http.go

```diff
@@ -422,109 +376,269 @@ func (s *httpService) getBackendSticky(req *http.Request) (*httputil.ClientConn,
 		}
 	}
 	if !ok {
-		return s.getNewBackendSticky()
+		return ""
 	}
 
-	backend, err := net.Dial("tcp", string(addr))
+	return addr
+}
+
+func (s *httpService) newStickyCookie(backend string) *http.Cookie {
+	var nonce [24]byte
+	_, err := io.ReadFull(rand.Reader, nonce[:])
 	if err != nil {
-		return s.getNewBackendSticky()
+		panic(err)
 	}
-	return httputil.NewClientConn(backend, nil), nil
+	out := make([]byte, len(nonce), len(nonce)+len(backend)+secretbox.Overhead)
+	copy(out, nonce[:])
+	out = secretbox.Seal(out, []byte(backend), &nonce, s.cookieKey)
+
+	return &http.Cookie{Name: stickyCookie, Value: base64.StdEncoding.EncodeToString(out), Path: "/"}
 }
 
-func (s *httpService) handle(req *http.Request, sc *httputil.ServerConn, tls, sticky bool) (done bool) {
+func (s *httpService) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	req.Header.Set("X-Request-Start", strconv.FormatInt(time.Now().UnixNano()/int64(time.Millisecond), 10))
 	req.Header.Set("X-Request-Id", random.UUID())
 
-	var backend *httputil.ClientConn
-	var stickyCookie *http.Cookie
-	if sticky {
-		backend, stickyCookie = s.getBackendSticky(req)
-	} else {
-		backend = s.getBackend()
-	}
-	if backend == nil {
-		log.Println("no backend found")
-		fail(sc, req, 503, "Service Unavailable")
+	addrs := shuffle(s.ss.Addrs())
+	if len(addrs) == 0 {
+		log.Println("no backends found")
+		fail(w, 503, "Service Unavailable")
 		return
 	}
-	defer backend.Close()
-
-	req.Proto = "HTTP/1.1"
-	req.ProtoMajor = 1
-	req.ProtoMinor = 1
-	delete(req.Header, "Te")
-	delete(req.Header, "Transfer-Encoding")
 
-	if clientIP, _, err := net.SplitHostPort(req.RemoteAddr); err == nil {
-		// If we aren't the first proxy retain prior
-		// X-Forwarded-For information as a comma+space
-		// separated list and fold multiple headers into one.
-		if prior, ok := req.Header["X-Forwarded-For"]; ok {
-			clientIP = strings.Join(prior, ", ") + ", " + clientIP
+	isSticky := false
+	stickyAddr := ""
+	if req.Header.Get(hdrUseStickySessions) == "true" {
+		// TODO(bgentry): switch to better way to check sticky setting
+		isSticky = true
+		req.Header.Del(hdrUseStickySessions)
+		if stickyAddr = s.stickyCookieAddr(req); stickyAddr != "" {
+			sortStringFirst(addrs, stickyAddr)
 		}
-		req.Header.Set("X-Forwarded-For", clientIP)
 	}
-	if tls {
-		req.Header.Set("X-Forwarded-Proto", "https")
-	} else {
-		req.Header.Set("X-Forwarded-Proto", "http")
-	}
-	// TODO: Set X-Forwarded-Port
+
+	// Most of this is borrowed from httputil.ReverseProxy
+	outreq := new(http.Request)
```

Comment:
We've standardized on not using `new` to initialize structs, use `&http.Request{}`.